pkg/apis/certmanager/validation/issuer_test.go

```diff
@@ -0,0 +1,485 @@
+			},
+		},
+		"missing akamai config": {
+			cfg: &v1alpha1.ACMEIssuerDNS01Config{
+				Providers: []v1alpha1.ACMEIssuerDNS01Provider{
+					{
+						Name:   "a name",
+						Akamai: &v1alpha1.ACMEIssuerDNS01ProviderAkamai{},
+					},
+				},
+			},
+			errs: []*field.Error{
+				field.Required(providersPath.Index(0).Child("akamai", "accessToken", "name"), "secret name is required"),
+				field.Required(providersPath.Index(0).Child("akamai", "accessToken", "key"), "secret key is required"),
+				field.Required(providersPath.Index(0).Child("akamai", "clientSecret", "name"), "secret name is required"),
+				field.Required(providersPath.Index(0).Child("akamai", "clientSecret", "key"), "secret key is required"),
+				field.Required(providersPath.Index(0).Child("akamai", "clientToken", "name"), "secret name is required"),
+				field.Required(providersPath.Index(0).Child("akamai", "clientToken", "key"), "secret key is required"),
+				field.Required(providersPath.Index(0).Child("akamai", "serviceConsumerDomain"), ""),
+			},
+		},
+		"valid akamai config": {
+			cfg: &v1alpha1.ACMEIssuerDNS01Config{
+				Providers: []v1alpha1.ACMEIssuerDNS01Provider{
+					{
+						Name: "a name",
+						Akamai: &v1alpha1.ACMEIssuerDNS01ProviderAkamai{
+							AccessToken:           validSecretKeyRef,
+							ClientSecret:          validSecretKeyRef,
+							ClientToken:           validSecretKeyRef,
+							ServiceConsumerDomain: "abc",
+						},
+					},
+				},
+			},
+			errs: []*field.Error{},
+		},
+		"multiple providers configured": {
+			cfg: &v1alpha1.ACMEIssuerDNS01Config{
+				Providers: []v1alpha1.ACMEIssuerDNS01Provider{
+					{
+						Name:       "a name",
+						CloudDNS:   &validCloudDNSProvider,
+						Cloudflare: &validCloudflareProvider,
+					},
+				},
+			},
+			errs: []*field.Error{
+				field.Forbidden(providersPath.Index(0).Child("cloudflare"), "may not specify more than one provider type"),
+			},
+		},
+	}
+	for n, s := range scenarios {
+		t.Run(n, func(t *testing.T) {
+			errs := ValidateACMEIssuerDNS01Config(s.cfg, fldPath)
+			if len(errs) != len(s.errs) {
+				t.Errorf("Expected %v but got %v", s.errs, errs)
+				return
+			}
+			for i, e := range errs {
+				expectedErr := s.errs[i]
+				if !reflect.DeepEqual(e, expectedErr) {
+					t.Errorf("Expected %v but got %v", expectedErr, e)
+				}
+			}
+		})
+	}
+}
+
+func TestValidateSecretKeySelector(t *testing.T) {
+	validName := v1alpha1.LocalObjectReference{"name"}
+	validKey := "key"
+	// invalidName := v1alpha1.LocalObjectReference{"-name-"}
+	// invalidKey := "-key-"
+	fldPath := field.NewPath("")
+	scenarios := map[string]struct {
+		isExpectedFailure bool
+		selector          *v1alpha1.SecretKeySelector
+		errs              []*field.Error
+	}{
+		"valid selector": {
+			selector: &v1alpha1.SecretKeySelector{
+				LocalObjectReference: validName,
+				Key:                  validKey,
+			},
+		},
+		// "invalid name": {
```

Comment:
Because we don't currently fail validation in these cases, although we probably should. I'd written up the test cases before realising we don't gate on this. I suppose we are better to actually gate on it 😬 I need to double check what the actual requirements here should be however.